File: packages/clippster-editor-native/android/src/main/java/app/clippster/editor/engine/ClippsterEditorPreviewView.kt
package app.clippster.editor.engine

import android.graphics.Color
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.SurfaceTexture
import android.view.Surface
import android.view.TextureView
import android.widget.FrameLayout
import android.widget.ImageView
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import org.json.JSONObject

class ClippsterEditorPreviewView(
  context: android.content.Context,
  appContext: AppContext,
) : ExpoView(context, appContext), TextureView.SurfaceTextureListener {
  private val onSurfaceReady by EventDispatcher()
  private val onFramePresented by EventDispatcher()
  private val textureView = TextureView(context)
  private val bitmapPreviewView = ImageView(context)
  private val overlayView = LayerOverlayView(context)
  private val pipeline = MediaPreviewPipeline(context)
  private val bitmapPipeline = BitmapPreviewPipeline(context)
  private var previewBitmap: android.graphics.Bitmap? = null

  private var documentJson = ""
  private var playheadSeconds = 0.0
  private var playing = false
  private var quality = "auto"
  private var lastRequestedSeconds = Double.NaN
  private var lastRequestedSource: String? = null

  init {
    setBackgroundColor(Color.BLACK)
    // TextureView rejects background drawables on API 28+; keep black on the parent only.
    textureView.surfaceTextureListener = this
    addView(
      textureView,
      FrameLayout.LayoutParams(
        FrameLayout.LayoutParams.MATCH_PARENT,
        FrameLayout.LayoutParams.MATCH_PARENT,
      ),
    )
    bitmapPreviewView.scaleType = ImageView.ScaleType.FIT_XY
    bitmapPreviewView.visibility = INVISIBLE
    addView(
      bitmapPreviewView,
      FrameLayout.LayoutParams(
        FrameLayout.LayoutParams.MATCH_PARENT,
        FrameLayout.LayoutParams.MATCH_PARENT,
      ),
    )
    addView(
      overlayView,
      FrameLayout.LayoutParams(
        FrameLayout.LayoutParams.MATCH_PARENT,
        FrameLayout.LayoutParams.MATCH_PARENT,
      ),
    )
  }

  fun setDocumentJson(value: String) {
    if (documentJson == value) return
    documentJson = value
    lastRequestedSeconds = Double.NaN
    lastRequestedSource = null
    requestFrame()
  }

  fun setPlaying(value: Boolean) {
    playing = value
    if (!value) {
      lastRequestedSeconds = Double.NaN
    }
  }

  fun setPlayheadSeconds(value: Double) {
    playheadSeconds = value.coerceAtLeast(0.0)
    requestFrame()
  }

  fun setQuality(value: String) {
    if (quality == value) return
    quality = value
    lastRequestedSeconds = Double.NaN
    requestFrame()
  }

  private fun requestFrame() {
    if (documentJson.isBlank()) return
    val frame = evaluatedFrame()
    overlayView.setFrame(frame)
    if (!textureView.isAvailable) return
    val videoLayer = frame?.optJSONArray("layers")
      ?.let { layers ->
        (0 until layers.length())
          .mapNotNull(layers::optJSONObject)
          .firstOrNull { it.optString("kind") == "video" }
      }
    val videoLayers = frame?.optJSONArray("layers")?.let { layers ->
      (0 until layers.length())
        .mapNotNull(layers::optJSONObject)
        .filter { it.optString("kind") == "video" }
    }.orEmpty()
    val needsBitmapComposition = videoLayers.any(::requiresBitmapComposition)
    if (needsBitmapComposition && frame != null && width > 0 && height > 0) {
      textureView.visibility = INVISIBLE
      bitmapPreviewView.visibility = VISIBLE
      bitmapPipeline.render(frame, width, height) { bitmap ->
        post {
          val previous = previewBitmap
          previewBitmap = bitmap
          bitmapPreviewView.setImageBitmap(bitmap)
          previous?.takeIf { it !== bitmap && !it.isRecycled }?.recycle()
          onFramePresented(mapOf("timeSeconds" to playheadSeconds, "quality" to quality))
        }
      }
      return
    }
    bitmapPipeline.cancel()
    bitmapPreviewView.visibility = INVISIBLE
    textureView.visibility = VISIBLE
    applyEvaluatorTransform(frame, videoLayer)
    val source = videoLayer?.optString("sourceUri")?.takeIf(String::isNotBlank)
      ?: (if (frame == null) firstVideoSource(documentJson) else null)
      ?: return
    val sourceTick = videoLayer
      ?.takeIf { it.has("sourceTick") && !it.isNull("sourceTick") }
      ?.optLong("sourceTick")
    val decodeTimeSeconds = sourceTick
      ?.let { it.toDouble() / ClippsterEditorNativeModule.ticksPerSecond() }
      ?: playheadSeconds

    // Drop redundant play ticks — sticky decoder advances; flooding seeks stalls it.
    if (
      playing &&
      source == lastRequestedSource &&
      !lastRequestedSeconds.isNaN() &&
      kotlin.math.abs(decodeTimeSeconds - lastRequestedSeconds) < PLAY_TICK_MIN_DELTA
    ) {
      return
    }
    lastRequestedSource = source
    lastRequestedSeconds = decodeTimeSeconds

    pipeline.showFrame(
      sourceUri = source,
      timeSeconds = decodeTimeSeconds,
      precise = !playing,
      onPresented = { seconds ->
        post {
          onFramePresented(
            mapOf(
              "timeSeconds" to seconds,
              "quality" to quality,
            ),
          )
        }
      },
      onError = { error ->
        post {
          // Missing/offline media is a recoverable black-frame state.
          onFramePresented(
            mapOf(
              "timeSeconds" to playheadSeconds,
              "error" to (error.message ?: "Unable to decode preview frame"),
            ),
          )
        }
      },
    )
  }

  private fun requiresBitmapComposition(layer: JSONObject): Boolean {
    if (layer.optJSONObject("transition")?.optString("kind") == "wipe") return true
    val effects = layer.optJSONArray("effects") ?: return false
    for (index in 0 until effects.length()) {
      val type = effects.optJSONObject(index)?.optString("type")
        ?.lowercase()
        ?.replace('-', '_')
        ?.removePrefix("adjust_")
        ?: continue
      if (type in RASTER_EFFECTS) return true
    }
    return false
  }

  private fun firstVideoSource(json: String): String? = runCatching {
    val root = JSONObject(json)
    val assets = root.optJSONObject("assets") ?: return@runCatching null
    assets.keys().asSequence()
      .mapNotNull { assets.optJSONObject(it) }
      .firstOrNull { it.optString("kind", "video").equals("video", true) }
      ?.optString("sourceUri")
      ?.takeIf(String::isNotBlank)
  }.getOrNull()

  private fun evaluatedFrame(): JSONObject? = runCatching {
    val tick = (playheadSeconds * ClippsterEditorNativeModule.ticksPerSecond()).toLong()
    JSONObject(
      ClippsterEditorNativeModule.evaluateDocument(
        documentJson,
        tick,
        true,
      ),
    ).takeUnless { it.has("error") }
  }.getOrNull()

  private fun applyEvaluatorTransform(frame: JSONObject?, layer: JSONObject?) {
    val frameCanvas = frame?.optJSONObject("canvas")
    val transform = layer?.optJSONObject("transform")
    if (frameCanvas == null || transform == null || width <= 0 || height <= 0) {
      textureView.setTransform(Matrix())
      textureView.alpha = 1f
      textureView.clipBounds = null
      textureView.setLayerType(LAYER_TYPE_NONE, null)
      return
    }
    textureView.alpha = layer.optDouble("opacity", 1.0).coerceIn(0.0, 1.0).toFloat()
    val colorFilter = EffectColorMatrices.colorFilter(layer.optJSONArray("effects"))
    textureView.setLayerType(
      if (colorFilter == null) LAYER_TYPE_NONE else LAYER_TYPE_HARDWARE,
      colorFilter?.let { Paint(Paint.ANTI_ALIAS_FLAG).apply { this.colorFilter = it } },
    )
    val transition = layer.optJSONObject("transition")
    textureView.clipBounds =
      if (transition?.optString("kind") == "wipe") {
        val progress = transition.optDouble("horizontalWipeProgress", 1.0)
          .coerceIn(0.0, 1.0)
        Rect(0, 0, (textureView.width * progress).toInt(), textureView.height)
      } else {
        null
      }
    val canvasWidth = frameCanvas.optDouble("width", width.toDouble()).coerceAtLeast(1.0)
    val canvasHeight = frameCanvas.optDouble("height", height.toDouble()).coerceAtLeast(1.0)
    val sourceWidth = layer.optDouble("sourceWidth", width.toDouble()).coerceAtLeast(1.0)
    val sourceHeight = layer.optDouble("sourceHeight", height.toDouble()).coerceAtLeast(1.0)
    val outputScaleX = width / canvasWidth
    val outputScaleY = height / canvasHeight
    val inputScaleX = sourceWidth / width
    val inputScaleY = sourceHeight / height
    val values = floatArrayOf(
      (outputScaleX * transform.optDouble("m00", 1.0) * inputScaleX).toFloat(),
      (outputScaleX * transform.optDouble("m01", 0.0) * inputScaleY).toFloat(),
      (outputScaleX * transform.optDouble("m02", 0.0)).toFloat(),
      (outputScaleY * transform.optDouble("m10", 0.0) * inputScaleX).toFloat(),
      (outputScaleY * transform.optDouble("m11", 1.0) * inputScaleY).toFloat(),
      (outputScaleY * transform.optDouble("m12", 0.0)).toFloat(),
      0f,
      0f,
      1f,
    )
    textureView.setTransform(Matrix().apply { setValues(values) })
  }

  override fun onSurfaceTextureAvailable(surfaceTexture: SurfaceTexture, width: Int, height: Int) {
    pipeline.attachSurface(Surface(surfaceTexture))
    onSurfaceReady(mapOf("width" to width, "height" to height))
    requestFrame()
  }

  override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture, width: Int, height: Int) {
    requestFrame()
  }

  override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
    pipeline.detachSurface()
    return true
  }

  override fun onSurfaceTextureUpdated(surface: SurfaceTexture) = Unit

  override fun onDetachedFromWindow() {
    pipeline.release()
    bitmapPipeline.release()
    bitmapPreviewView.setImageDrawable(null)
    if (previewBitmap?.isRecycled == false) previewBitmap?.recycle()
    previewBitmap = null
    super.onDetachedFromWindow()
  }

  private companion object {
    const val PLAY_TICK_MIN_DELTA = 1.0 / 45.0
    val RASTER_EFFECTS = setOf(
      "blur",
      "glitch",
      "letterbox",
      "vignette",
      "grain",
      "mirror",
    )
  }
}

File: packages/clippster-editor-native/ios/ClippsterEditorPreviewView.swift
import AVFoundation
import ExpoModulesCore
import UIKit

final class ClippsterEditorPreviewView: ExpoView {
  let onSurfaceReady = EventDispatcher()
  let onFramePresented = EventDispatcher()

  private let displayLayer = AVSampleBufferDisplayLayer()
  private let filteredFrameView = UIImageView()
  private let overlayView = LayerOverlayView()
  private let pipeline = MediaPreviewPipeline()
  private let wipeRenderer = WipePreviewRenderer()
  private var documentJSON = ""
  private var frameDescriptor: [String: Any]?
  private var playheadSeconds = 0.0
  private var playing = false
  private var quality = "auto"
  private var showingWipeComposite = false
  private var lastRequestedSeconds = Double.nan
  private var lastRequestedSource: String?

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    backgroundColor = .black
    displayLayer.backgroundColor = UIColor.black.cgColor
    displayLayer.videoGravity = .resize
    displayLayer.anchorPoint = .zero
    layer.addSublayer(displayLayer)
    filteredFrameView.isHidden = true
    filteredFrameView.contentMode = .scaleToFill
    filteredFrameView.layer.anchorPoint = .zero
    addSubview(filteredFrameView)
    addSubview(overlayView)
    pipeline.attach(to: displayLayer)
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    if showingWipeComposite {
      filteredFrameView.frame = bounds
    } else {
      applyEvaluatorTransform(frameDescriptor)
    }
    overlayView.frame = bounds
    onSurfaceReady(["width": bounds.width, "height": bounds.height])
  }

  func setDocumentJSON(_ value: String) {
    if documentJSON == value { return }
    documentJSON = value
    lastRequestedSeconds = .nan
    lastRequestedSource = nil
    requestFrame()
  }

  func setPlaying(_ value: Bool) {
    playing = value
    if !value {
      lastRequestedSeconds = .nan
    }
  }

  func setPlayheadSeconds(_ value: Double) {
    playheadSeconds = max(0, value)
    requestFrame()
  }

  func setQuality(_ value: String) {
    if quality == value { return }
    quality = value
    lastRequestedSeconds = .nan
    requestFrame()
  }

  private func requestFrame() {
    let frame = evaluatedFrame()
    frameDescriptor = frame
    overlayView.setFrame(frame)
    let videoLayers = (frame?["layers"] as? [[String: Any]])?
      .filter { $0["kind"] as? String == "video" } ?? []
    let videoLayer = videoLayers.first
    applyEvaluatorTransform(frame)
    let needsWipeFallback = videoLayers.contains {
      ($0["transition"] as? [String: Any])?["kind"] as? String == "wipe"
    }
    if needsWipeFallback, let frame, bounds.width > 0, bounds.height > 0 {
      showingWipeComposite = true
      displayLayer.isHidden = true
      filteredFrameView.isHidden = false
      filteredFrameView.layer.setAffineTransform(.identity)
      filteredFrameView.layer.mask = nil
      filteredFrameView.frame = bounds
      wipeRenderer.render(frame: frame, outputSize: bounds.size) { [weak self] image in
        guard let self else { return }
        filteredFrameView.image = image
        onFramePresented(["timeSeconds": playheadSeconds, "quality": quality])
      }
      return
    }
    wipeRenderer.cancel()
    showingWipeComposite = false
    displayLayer.isHidden = false
    let source = videoLayer?["sourceUri"] as? String
      ?? (frame == nil ? firstVideoSource() : nil)
    guard window != nil, let source, !source.isEmpty else { return }
    let sourceTick = (videoLayer?["sourceTick"] as? NSNumber)?.int64Value
    let decodeSeconds = sourceTick.map {
      Double($0) / Double(GraphBridge.ticksPerSecond())
    } ?? playheadSeconds

    if playing,
       source == lastRequestedSource,
       !lastRequestedSeconds.isNaN,
       abs(decodeSeconds - lastRequestedSeconds) < (1.0 / 45.0) {
      return
    }
    lastRequestedSource = source
    lastRequestedSeconds = decodeSeconds

    let effects = videoLayer?["effects"] as? [[String: Any]] ?? []
    let clipID = videoLayer?["clipId"] as? String ?? ""
    let frameTick = (frame?["tick"] as? NSNumber)?.int64Value ?? sourceTick ?? 0
    let effectSeed = EffectImageProcessor.stableSeed(clipID, tick: frameTick)
    displayLayer.isHidden = !effects.isEmpty
    filteredFrameView.isHidden = effects.isEmpty
    let frameHandler: ((UIImage) -> Void)? = effects.isEmpty ? nil : { [weak self] image in
      self?.filteredFrameView.image = EffectImageProcessor.process(
        image,
        effects: effects,
        seed: effectSeed
      )
    }
    pipeline.showFrame(
      sourceURI: source,
      seconds: decodeSeconds,
      precise: !playing,
      frameHandler: frameHandler
    ) { [weak self] result in
      guard let self else { return }
      switch result {
      case .success(let seconds):
        onFramePresented(["timeSeconds": seconds, "quality": quality])
      case .failure(let error):
        // Offline/missing media is a recoverable black-frame state.
        displayLayer.flushAndRemoveImage()
        onFramePresented([
          "timeSeconds": playheadSeconds,
          "error": error.localizedDescription,
        ])
      }
    }
  }

  private func firstVideoSource() -> String? {
    guard
      let data = documentJSON.data(using: .utf8),
      let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
      let assets = root["assets"] as? [String: Any]
    else { return nil }

    return assets.values.lazy.compactMap { value -> String? in
      guard let asset = value as? [String: Any] else { return nil }
      let kind = (asset["kind"] as? String ?? "video").lowercased()
      return kind == "video" ? asset["sourceUri"] as? String : nil
    }.first
  }

  private func evaluatedFrame() -> [String: Any]? {
    let tick = Int64(playheadSeconds * Double(GraphBridge.ticksPerSecond()))
    guard
      let json = try? GraphBridge.parseAndEvaluate(
        documentJSON,
        tick: tick,
        previewMode: true
      ),
      let data = json.data(using: .utf8)
    else { return nil }
    return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
  }

  private func applyEvaluatorTransform(_ frame: [String: Any]?) {
    guard
      let layers = frame?["layers"] as? [[String: Any]],
      let layer = layers.first(where: { $0["kind"] as? String == "video" }),
      let canvas = frame?["canvas"] as? [String: Any],
      let canvasWidth = number(canvas["width"]),
      let canvasHeight = number(canvas["height"]),
      let sourceWidth = number(layer["sourceWidth"]),
      let sourceHeight = number(layer["sourceHeight"]),
      canvasWidth > 0,
      canvasHeight > 0,
      sourceWidth > 0,
      sourceHeight > 0
    else {
      displayLayer.setAffineTransform(.identity)
      displayLayer.bounds = bounds
      displayLayer.position = .zero
      displayLayer.opacity = 1
      displayLayer.mask = nil
      filteredFrameView.layer.setAffineTransform(.identity)
      filteredFrameView.frame = bounds
      filteredFrameView.layer.opacity = 1
      filteredFrameView.layer.mask = nil
      return
    }
    let outputScaleX = bounds.width / canvasWidth
    let outputScaleY = bounds.height / canvasHeight
    displayLayer.bounds = CGRect(x: 0, y: 0, width: sourceWidth, height: sourceHeight)
    displayLayer.position = .zero
    displayLayer.opacity = Float(
      min(max(number(layer["opacity"]) ?? 1, 0), 1)
    )
    let transform = OverlayFrameRenderer.affineTransform(
      for: layer,
      outputScaleX: outputScaleX,
      outputScaleY: outputScaleY
    )
    displayLayer.setAffineTransform(transform)
    filteredFrameView.layer.bounds = displayLayer.bounds
    filteredFrameView.layer.position = .zero
    filteredFrameView.layer.opacity = displayLayer.opacity
    filteredFrameView.layer.setAffineTransform(transform)
    let transition = layer["transition"] as? [String: Any]
    if transition?["kind"] as? String == "wipe" {
      let progress = min(max(number(transition?["horizontalWipeProgress"]) ?? 1, 0), 1)
      let mask = CALayer()
      mask.frame = CGRect(
        x: 0,
        y: 0,
        width: sourceWidth * progress,
        height: sourceHeight
      )
      mask.backgroundColor = UIColor.white.cgColor
      displayLayer.mask = mask
      let imageMask = CALayer()
      imageMask.frame = mask.frame
      imageMask.backgroundColor = UIColor.white.cgColor
      filteredFrameView.layer.mask = imageMask
    } else {
      displayLayer.mask = nil
      filteredFrameView.layer.mask = nil
    }
  }

  private func number(_ value: Any?) -> CGFloat? {
    (value as? NSNumber).map { CGFloat(truncating: $0) }
  }

  deinit {
    pipeline.release()
    wipeRenderer.release()
  }
}
